client/core/core.go

```diff
@@ -0,0 +1,251 @@
+	// enables creating custom loggers for use in a GUI interface.
+	Logger dex.Logger
+	// Certs is a mapping of URL to filepaths of TLS Certificates for the server.
+	// This is intended for accomodating self-signed certificates, mostly for
+	// testing.
+	Certs map[string]string
+}
+
+// Core is the core client application.
+type Core struct {
+	ctx     context.Context
+	wg      sync.WaitGroup
+	cfg     *Config
+	connMtx sync.RWMutex
+	conns   map[string]*dexConnection
+	db      db.DB
+	certs   map[string]string
+}
+
+// New is the constructor for a new Core.
+func New(cfg *Config) *Core {
+	log = cfg.Logger
+	core := &Core{
+		cfg:   cfg,
+		conns: make(map[string]*dexConnection),
+	}
+	return core
+}
+
+// Run runs the core. Satisfies the runner.Runner interface.
+func (c *Core) Run(ctx context.Context) {
+	log.Infof("started DEX client core")
+	db, err := bolt.NewDB(ctx, c.cfg.DBPath)
+	if err != nil {
+		log.Errorf("database initialization error: %v", err)
+		return
+	}
+	c.ctx = ctx
+	c.db = db
+	go c.initialize()
+	c.wg.Wait()
+	log.Infof("DEX client core off")
+}
+
+// ListMarkets returns a list of known markets.
+func (c *Core) ListMarkets() []*MarketInfo {
+	c.connMtx.RLock()
+	defer c.connMtx.RUnlock()
+	var infos []*MarketInfo
+	for uri, dc := range c.conns {
+		mi := &MarketInfo{DEX: uri}
+		for _, bq := range dc.cfg.Markets {
+			base, quote := dc.assets[bq[0]], dc.assets[bq[1]]
+			mi.Markets = append(mi.Markets, Market{
+				BaseID:      base.ID,
+				BaseSymbol:  base.Symbol,
+				QuoteID:     quote.ID,
+				QuoteSymbol: quote.Symbol,
+			})
+		}
+		infos = append(infos, mi)
+	}
+	return infos
+}
+
+// initialize pulls the known DEX URLs from the database and attempts to
+// connect and retreive the DEX configuration.
+func (c *Core) initialize() {
+	dexs, err := c.db.ListAccounts()
+	if err != nil {
+		log.Errorf("Error retreiving accounts from database: %v", err)
+	}
+	var wg sync.WaitGroup
+	for _, uri := range dexs {
+		wg.Add(1)
+		u := uri
+		go func() {
+			c.addDex(u)
+			wg.Done()
+		}()
+	}
+	wg.Wait()
+	if len(dexs) > 0 {
+		c.connMtx.RLock()
+		log.Infof("Successfully connected to %d out of %d DEX servers", len(c.conns), len(dexs))
+		c.connMtx.RUnlock()
+	}
+}
+
+// addDex adds a dexConnection to the conns map if a connection can be made
+// and the DEX configuration is successfully retrieved. The connection is
+// unauthenticated until the `connect` request is sent and accepted by the
+// server.
+func (c *Core) addDex(uri string) {
+	// Get the host from the DEX URL.
+	parsedURL, err := url.Parse(uri)
+	if err != nil {
+		log.Errorf("error parsing account URL %s: %v", uri, err)
+		return
+	}
+	// Create a websocket connection to the server.
+	conn, err := comms.NewWsConn(&comms.WsCfg{
+		URL:      "wss://" + parsedURL.Host + "/ws",
+		PingWait: 60 * time.Second,
+		RpcCert:  c.certs[uri],
+		ReconnectSync: func() {
+			go c.handleReconnect(uri)
+		},
+		Ctx: c.ctx,
```

Comment:
We've gotten even more pushback about this, but we can continue to feel it out.  No need to change for the unexported methods of `Core` I say.

Reply:
This pattern was in *client/comms* before, just hasn't been updated yet. For *core.Core*, I've used the `runner.Runner` interface.